storage/innobase/xtrabackup/src/ds_fifo.cc

```diff
@@ -0,0 +1,170 @@
+static int fifo_close(ds_file_t *file);
+static void fifo_deinit(ds_ctxt_t *ctxt);
+
+datasink_t datasink_fifo = {&fifo_init, &fifo_open,  &fifo_write,
+                            nullptr,    &fifo_close, &fifo_deinit};
+
+static ds_ctxt_t *fifo_init(const char *root) {
+  ds_fifo_ctxt_t *fifo_context = new ds_fifo_ctxt_t;
+  File fd;
+  ds_ctxt_t *ctxt = new ds_ctxt_t;
+  char fullpath[FN_REFLEN];
+
+  if (pthread_mutex_init(&fifo_context->fifo_mutex, NULL)) {
+    msg("fifo_init: pthread_mutex_init() failed.\n");
+    return NULL;
+  }
+
+  if (my_mkdir(root, 0777, MYF(0)) < 0 && my_errno() != EEXIST &&
+      my_errno() != EISDIR) {
+    char errbuf[MYSYS_STRERROR_SIZE];
+    my_error(EE_CANT_MKDIR, MYF(ME_BELL), root, my_errno(),
+             my_strerror(errbuf, sizeof(errbuf), my_errno()));
+    return NULL;
+  }
+
+  for (uint i = 0; i < xtrabackup_parallel_stream; i++) {
+    std::string path = "thread_" + std::to_string(i);
+    fn_format(fullpath, path.c_str(), root, "", MYF(MY_RELATIVE_PATH));
+    if (mkfifo(fullpath, 0666) < 0) {
+      msg("mkfifo(%s) failed with error %d", fullpath, errno);
+      return NULL;
+    }
+  }
+
+  for (uint i = 0; i < xtrabackup_parallel_stream; i++) {
+    std::string path = "thread_" + std::to_string(i);
+    fn_format(fullpath, path.c_str(), root, "", MYF(MY_RELATIVE_PATH));
+    fd = open_fifo_for_write_with_timeout(fullpath, xtrabackup_stream_timeout);
+    if (fd < 0) return NULL;
+
+    fifo_context->FIFO_list.insert({fullpath, fd});
+  }
+
+  ctxt->ptr = fifo_context;
+  ctxt->root = my_strdup(PSI_NOT_INSTRUMENTED, root, MYF(MY_FAE));
+
+  return ctxt;
+}
+
+static ds_file_t *fifo_open(ds_ctxt_t *ctxt,
+                            const char *path __attribute__((unused)),
+                            MY_STAT *mystat __attribute__((unused))) {
+  char fullpath[FN_REFLEN];
+  size_t path_len;
+  ds_fifo_file_t *fifo_file;
+  ds_file_t *file;
+  ds_fifo_ctxt_t *fifo_context = (ds_fifo_ctxt_t *)ctxt->ptr;
+  File fd;
+  std::string fifo_path;
+  pthread_mutex_lock(&fifo_context->fifo_mutex);
+  auto fifo_it = fifo_context->FIFO_list.begin();
+  if (fifo_it == fifo_context->FIFO_list.end()) {
+    pthread_mutex_unlock(&fifo_context->fifo_mutex);
+    return NULL;
+  }
+  fifo_path = fifo_it->first;
+  fd = fifo_it->second;
+  fn_format(fullpath, fifo_path.c_str(), ctxt->root, "", MYF(MY_RELATIVE_PATH));
+  fifo_context->FIFO_list.erase(fifo_it);
+  pthread_mutex_unlock(&fifo_context->fifo_mutex);
+
+  path_len = strlen(fullpath) + 1; /* terminating '\0' */
+
+  file = (ds_file_t *)my_malloc(
+      PSI_NOT_INSTRUMENTED,
+      sizeof(ds_file_t) + sizeof(ds_fifo_file_t) + path_len, MYF(MY_FAE));
+  fifo_file = (ds_fifo_file_t *)(file + 1);
+
+  fifo_file->fd = fd;
+  fifo_file->fifo_context = fifo_context;
+  fifo_file->path = (char *)fifo_path.c_str();
+  file->path = (char *)fifo_file + sizeof(ds_fifo_file_t);
+  memcpy(file->path, fullpath, path_len);
+
+  file->ptr = fifo_file;
```

Comment:
That is filled on the base ds_open:

```
/************************************************************************
Open a datasink file */
ds_file_t *ds_open(ds_ctxt_t *ctxt, const char *path, MY_STAT *stat) {
  ds_file_t *file;

  file = ctxt->datasink->open(ctxt, path, stat);
  if (file != NULL) {
    file->datasink = ctxt->datasink;
  }

  return file;
}
```

` file = ctxt->datasink->open(ctxt, path, stat);` this will call `fifo_open`, in case we return a file that is not null, file will inherit the datasink from the context.